modules/history/src/views/full/MessageViewer.jsx

```diff
@@ -164,19 +185,38 @@ export class MessageViewer extends React.Component {
             this.state.inspectorIsShown ? style['message-list-partial'] : style['message-list-full']
           )}
         >
-          <MessagesHeader conversation={this.props.conversation} messageGroups={this.props.messageGroups} />
-          {this.props.messageGroups &&
-            this.props.messageGroups.map(group => {
-              return (
-                <MessageGroup
-                  key={group[0].id}
-                  messages={group}
-                  focusMessage={m => this.setState({ currentlyFocusedMessage: m, inspectorIsShown: true })}
-                />
-              )
-            })}
+          <MessagesHeader conversation={this.state.currentConversation} messageGroups={this.props.messageGroups} />
+          <MessagesTaskBar
+            ref="taskBar"
+            useAsFilter={!this.state.areMessagesSelected}
+            flag={() => this.flagSelectedMessages()}
+            unflag={() => this.unflagSelectedMessages()}
+            updateFilters={f => this.updateFilters(f)}
+            currentConv={this.state.currentConversation}
+          />
+          {!!this.props.messageGroups.length && (
+            <div>
+              select all:
+              <input
+                type="checkbox"
+                checked={this.state.areAllMessagesSelected}
+                onChange={() => this.handleSelectAll()}
+              />
+              {this.props.messageGroups.map(group => {
+                return (
+                  <MessageGroup
+                    key={group.userMessage.id}
+                    group={group}
+                    focusMessage={m => this.setState({ currentlyFocusedMessage: m, inspectorIsShown: true })}
```

Comment:
You can avoid single chars when unnecessary. I would also suggest shortening the name, because there's only one focused message at a time

```js
focusMessage={focusedMessage => this.setState({ focusedMessage, inspectorIsShown: true })}
```